File: samples/declarative/assistants/virtual-assistant/Services/BotServices.cs
// Copyright (c) Microsoft Corporation. All rights reserved.
// Licensed under the MIT License.

using System;
using System.Collections.Generic;
using System.Globalization;
using System.Linq;
using Microsoft.Bot.Builder;
using Microsoft.Bot.Builder.AI.Luis;
using Microsoft.Bot.Builder.AI.QnA;
using Microsoft.Bot.Builder.Dialogs.Adaptive.QnA.Recognizers;
using Microsoft.Bot.Builder.Dialogs.Adaptive.Recognizers;
using Microsoft.Bot.Solutions;

namespace VirtualAssistantSample.Services
{
    public class BotServices
    {
        public BotServices()
        {
        }

        public BotServices(BotSettings settings, IBotTelemetryClient client)
        {
            foreach (var pair in settings.CognitiveModels)
            {
                var set = new AdaptiveCognitiveModelSet();
                var language = pair.Key;
                var config = pair.Value;

                var telemetryClient = client;

                LuisRecognizerOptionsV3 luisOptions;

                if (config.DispatchModel != null)
                {
                    var dispatchApp = new LuisAdaptiveRecognizer(){
                        ApplicationId = config.DispatchModel.AppId, 
                        EndpointKey = config.DispatchModel.SubscriptionKey,
                        Endpoint = config.DispatchModel.GetEndpoint(),
                        TelemetryClient = telemetryClient,
                        LogPersonalInformation = true,
                    };

                    set.DispatchService = dispatchApp;
                }

                if (config.LanguageModels != null)
                {
                    foreach (var model in config.LanguageModels)
                    {                        
                        var luisApp = new LuisAdaptiveRecognizer()
                        {
                            ApplicationId = model.AppId,
                            EndpointKey = model.SubscriptionKey,
                            Endpoint = model.GetEndpoint(),
                            TelemetryClient = telemetryClient,
                            LogPersonalInformation = true,
                        };
                        
                        set.LuisServices.Add(model.Id, luisApp);
                    }
                }

                foreach (var kb in config.Knowledgebases)
                {
                    var qnaEndpoint = new QnAMakerEndpoint()
                    {
                        KnowledgeBaseId = kb.KbId,
                        EndpointKey = kb.EndpointKey,
                        Host = kb.Hostname
                    };

                    set.QnAConfiguration.Add(kb.Id, qnaEndpoint);
                }

                CognitiveModelSets.Add(language, set);
            }
        }

        public Dictionary<string, AdaptiveCognitiveModelSet> CognitiveModelSets { get; set; } = new Dictionary<string, AdaptiveCognitiveModelSet>();

        public AdaptiveCognitiveModelSet GetCognitiveModels()
        {
            // Get cognitive models for locale
            var locale = CultureInfo.CurrentUICulture.Name.ToLower();

            var cognitiveModel = CognitiveModelSets.ContainsKey(locale)
                ? CognitiveModelSets[locale]
                : CognitiveModelSets.Where(key => key.Key.StartsWith(locale.Substring(0, 2))).FirstOrDefault().Value
                ?? throw new Exception($"There's no matching locale for '{locale}' or its root language '{locale.Substring(0, 2)}'. " +
                                        "Please review your available locales in your cognitivemodels.json file.");

            return cognitiveModel;
        }
    }
}

File: templates/typescript/samples/sample-assistant/src/services/botServices.ts
/**
 * Copyright(c) Microsoft Corporation.All rights reserved.
 * Licensed under the MIT License.
 */

import { BotTelemetryClient } from 'botbuilder';
import {
    LuisApplication,
    LuisRecognizer,
    LuisRecognizerOptionsV3,
    QnAMakerEndpoint } from 'botbuilder-ai';
import { CognitiveModelConfiguration, ICognitiveModelSet } from 'bot-solutions';
import { LuisService, QnaMakerService } from 'botframework-config';
import { IBotSettings } from '../services/botSettings';

export class BotServices {
    public cognitiveModelSets: Map<string, ICognitiveModelSet> = new Map();

    public constructor(settings: IBotSettings, client: BotTelemetryClient) {
        settings.cognitiveModels.forEach((value: CognitiveModelConfiguration, key: string): void => {
            const language: string = key;
            const config: CognitiveModelConfiguration = value;

            const telemetryClient: BotTelemetryClient = client;

            const luisOptions: LuisRecognizerOptionsV3 = {
                telemetryClient: telemetryClient,
                logPersonalInformation: true,
                apiVersion: 'v3'
            };

            const set: Partial<ICognitiveModelSet> = {
                luisServices: new Map(),
                qnaConfiguration: new Map(),
                qnaServices: new Map()
            };

            if (config.dispatchModel !== undefined) {
                const dispatchModel: LuisService = new LuisService(config.dispatchModel);
                const dispatchApp: LuisApplication = {
                    applicationId: config.dispatchModel.appId,
                    endpointKey: config.dispatchModel.subscriptionKey,
                    endpoint: dispatchModel.getEndpoint()
                };
                set.dispatchService= new LuisRecognizer(dispatchApp, luisOptions);
            }

            if (config.languageModels !== undefined) {
                config.languageModels.forEach((model: LuisService): void => {
                    const luisModel: LuisService = new LuisService(model);
                    const luisApp: LuisApplication  = {
                        applicationId: luisModel.appId,
                        endpointKey: luisModel.subscriptionKey,
                        endpoint: luisModel.getEndpoint()
                    };
                    if (set.luisServices !== undefined) {
                        set.luisServices.set(model.id, new LuisRecognizer(luisApp, luisOptions));
                    }
                });
            }

            config.knowledgeBases.forEach((kb: QnaMakerService): void => {
                const qnaEndpoint: QnAMakerEndpoint = {
                    knowledgeBaseId: kb.kbId,
                    endpointKey: kb.endpointKey,
                    host: kb.hostname
                };
                if (set.qnaConfiguration !== undefined) {
                    set.qnaConfiguration.set(kb.id, qnaEndpoint);
                }
            });

            this.cognitiveModelSets.set(language, set as ICognitiveModelSet);
        });
    }    

    public getCognitiveModels(locale: string): ICognitiveModelSet {
        // Get cognitive models for locale
        let cognitiveModels: ICognitiveModelSet | undefined = this.cognitiveModelSets.get(locale);

        if (cognitiveModels === undefined) {
            const keyFound: string | undefined = Array.from(this.cognitiveModelSets.keys())
                .find((key: string) => { key.substring(0, 2) === locale.substring(0, 2); });
            if (keyFound !== undefined) {
                cognitiveModels = this.cognitiveModelSets.get(keyFound);
            }
        }
        if (cognitiveModels === undefined) {
            throw new Error(`There's no matching locale for '${ locale }' or its root language '${ locale.substring(0, 2) }'.
            Please review your available locales in your cognitivemodels.json file.`);
        }

        return cognitiveModels;
    }
}
